### ISSUE TITLE
Add Union, Intersect, Except + Distinct

### ISSUE THREAD

**Issue by greystate** [greystate/umb-razor-reference#3]
In the **Filter & Order** category there's also these:

```
.Union(IEnumerable<IPublishedContent>[,IEqualityComparer])
.Intersect(IEnumerable<IPublishedContent>[,IEqualityComparer])
.Except(IEnumerable<IPublishedContent>[,IEqualityComparer])
.Distinct([IEqualityComparer])
```

But I'm not really down with that `IEqualityComparer` argument - is it a lambda Func, or how do I best describe it? I'd like to keep the descriptions and names "understandable" (e.g. intead of `IEnumerable<IPublishedContent>` I just say `collection` which is what you're really thinking :-)

**Comment by lars-erik** [comment]
It's an instance of a concrete IEqualityComparer.

**Comment by greystate** [comment]
Thanks Lars! (Still smoke and mirrors to me though...)

E.g.: If I were to use the `.Distinct()` method; do I only need to supply an `IEqualityComparer` if the collections aren't collections of same "type", or are there built-in comparers that can handle some cases?

Is it *almost* like a lambda Func?

**Comment by lars-erik** [comment]
It's basically like separating Object.Equals out into its own (polymorphic) class(es) for comparing things.
There are a few built-ins, but you're mostly supposed to make your own.
Say you'd like a distinct list of cities from form posts for instance. They might have different casing and might have trailing spaces. You could make an IEqualityComparer that ignores those differences. The distinct list would keep the first found of each city and ignore the matches.

I.E.
Oslo
oslo
Copenhagen
OSLO
copenhagen

Would yield

Oslo
Copenhagen

You could use the same equality comparer for a grouping clause that groups the posts by those "cleaned" distinct keys as well.

It's primary use-case is to tweak equality for things that aren't completely equal (or same reference).

Make sense?

**Comment by greystate** [comment]
Alright - yes, penny dropped (or whatever the proverb is :-) Thanks!

So what those methods are doing here (on "collections" of `IPublishedContent`) - the vast majority of uses would not need specifying one, I guess, since they (I pressume) already know how to tell if one node is equal to another?

**Comment by lars-erik** [comment]
Yes, and I can't really see many use-cases with published content. They're there in LINQ for everything and what-not so we can't really hide them.
